src/gfx_layout_icu.cpp

```diff
@@ -0,0 +1,311 @@
+ * See the GNU General Public License for more details. You should have received a copy of the GNU General Public License along with OpenTTD. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/** @file icu_layout.cpp Handling of laying out with ICU / Harfbuzz. */
+
+#include "stdafx.h"
+#include "gfx_layout_icu.h"
+
+#include "strings_func.h"
+#include "language.h"
+#include "3rdparty/icu/scriptrun.h"
+
+#include <unicode/ubidi.h>
+
+#include <hb.h>
+#include <hb-ft.h>
+
+#include <deque>
+
+#include "safeguards.h"
+
+/** harfbuzz doesn't use floats, so we need a value to scale position with to get sub-pixel precision. */
+constexpr int FONT_SCALE = 64.0;
+
+class Run {
+public:
+	int start;
+	int length;
+	UBiDiDirection direction;
+	UScriptCode script;
+	Font *font;
+
+	Run(int start, int length, UBiDiDirection direction, UScriptCode script = USCRIPT_UNKNOWN, Font *font = nullptr) : start(start), length(length), direction(direction), script(script), font(font) {}
+};
+
+/**
+ * Wrapper for doing layouts with ICU.
+ */
+class ICUParagraphLayout : public ParagraphLayouter {
+private:
+	bool show_line = false;
+
+public:
+	/** Visual run contains data about the bit of text with the same font. */
+	class ICUVisualRun : public ParagraphLayouter::VisualRun {
+	private:
+		std::vector<GlyphID> glyphs;
+		std::vector<float> positions;
+		const Font *font;
+		int total_advance;
+
+	public:
+		ICUVisualRun(UChar *buff, size_t length, Run &run, int x) : font(run.font) {
+			auto hbfont = hb_ft_font_create_referenced(*(static_cast<const FT_Face *>(font->fc->GetOSHandle())));
+
+			hb_font_set_scale(hbfont, this->font->fc->GetFontSize() * FONT_SCALE, this->font->fc->GetFontSize() * FONT_SCALE);
+
+			auto buf = hb_buffer_create();
+			hb_buffer_add_utf16(buf, reinterpret_cast<uint16 *>(buff), length, run.start, run.length);
+
+			hb_buffer_set_direction(buf, run.direction == UBIDI_RTL ? HB_DIRECTION_RTL : HB_DIRECTION_LTR);
+			hb_buffer_set_script(buf, hb_script_from_string(uscript_getShortName(run.script), -1));
+			hb_buffer_set_language(buf, hb_language_from_string(_current_language->isocode, -1));
+
+			hb_shape(hbfont, buf, nullptr, 0);
+
+			unsigned int glyph_count;
+			auto glyph_info = hb_buffer_get_glyph_infos(buf, &glyph_count);
+			auto glyph_pos = hb_buffer_get_glyph_positions(buf, &glyph_count);
+
+			hb_position_t advance = 0;
+			for (unsigned int i = 0; i < glyph_count; i++) {
+				this->glyphs.push_back(glyph_info[i].codepoint);
+				this->positions.push_back(glyph_pos[i].x_offset / FONT_SCALE + advance + x);
+				this->positions.push_back(glyph_pos[i].y_offset / FONT_SCALE);
+
+				advance += glyph_pos[i].x_advance / FONT_SCALE;
+			}
+
+			this->total_advance = advance;
+
+			this->positions.push_back(advance + x);
+			this->positions.push_back(0);
+
+			hb_buffer_destroy(buf);
+			hb_font_destroy(hbfont);
+		}
```

Comment:
Not sure I would have put the constructor inline, but that is purely a style choice.

Reply:
As I mentioned on Discord, this whole function needs to move to its own function anyway, so that will be fixed :) This was just me being lazy :D